### ISSUE TITLE
Support for Hugo Shortcodes in Rmd files

### ISSUE THREAD

**Issue by mwhitaker** [rstudio/blogdown#40]
Thank you for a great package! A really nice feature of Hugo is [shortcodes](https://gohugo.io/extras/shortcodes/). When Hugo encounters these snippets in a markdown file it renders small HTML templates into the output HTML file, eg

    {{< tweet 820477439050731521 >}}

would embed the tweet into the output HTML.

It would be great if support for shortcodes could be added to `Rmd` files. I tried to use `rmarkdown::md_document(preserve_yaml = TRUE)` to first render Rmd to md, but < and > in the shortcode snippet would always get rewritten as &lt; and &gt; in the rendered md file

    {{&lt; tweet 820477439050731521 &gt;}}

thereby thwarting Hugo.

Or is there perhaps a **knitr** option?

**Comment by yihui** [comment]
I'm aware of this issue, and my current guess is that we will have to reinvent some shortcodes using the R syntax, otherwise you have to make a hard decision on whether to give up shortcodes or give up R in your post... [It is not easy to have both](https://github.com/yihui/hugo-lithium-theme/blob/master/exampleSite/content/post/2016-12-30-hello-markdown.md) since they have different Markdown rendering engines.

**Comment by lcolladotor** [comment]
You can use `knitr::render_markdown(TRUE)` like I mention in #42. That's how I created https://github.com/lcolladotor/hugoblog/blob/master/content/post/2017-01-25-trumpwall.md which lead to https://github.com/lcolladotor/lcolladotor.github.com/blob/master/2017/01/25/An-alternative-to-the-Mexico-US-wall-where-the-US-would-gain-millions-of-dollars/index.html.

**Comment by yihui** [comment]
It turns out there is a much easier way:

    `r htmltools::HTML("{{< tweet 666616452582129664 >}}")`

This makes sure Pandoc does not touch the special characters`< >`.

**Comment by lcolladotor** [comment]
Awesome @yihui!

Great gif too hehe ^^

**Comment by yihui** [comment]
I'll provide a wrapper function in **blogdown** (something like `` `r blogdown::shortcode('tweet', '666616452582129664')` ``).

This problem was actually the most awkward drawback of using R Markdown on my mind when I started working on this project. I'm so glad that Hugo still processes shortcodes after I have converted (R) Markdown to HTML. Now I'm much more comfortable about R Markdown (other than TODO lists, pretty much Pandoc can do everything that Blackfriday can do).

**Comment by mwhitaker** [comment]
Eureka! Rendering shortcodes in Rmd files has definitely eluded me and I spent more than two hours on this ;-)

The nice thing is that you could now use this approach to save the data output for example from an htmlwidget into a JSON file into the static folder, and then pull in via a shortcode in a Hugo template using the `getJSON` Hugo [function](https://gohugo.io/extras/dynamiccontent/). So you could just render from Rmd to md and let Hugo do the rest. That works very well and I find that approach very elegant.

The fact that this also works in HTML is great.  Thank you!

**Comment by pssguy** [comment]
@mwhitaker Does that mean you could have a live twitter feed on a blogdown site?

**Comment by mwhitaker** [comment]
@pssguy sure. Would be quite easy to do as long as the live twitter feed uses a Javascript API, which I think does. You won't need the Hugo `getJSON` function I mentioned earlier, just set up a shortcode.

**Comment by mwhitaker** [comment]
@yihui I have one follow-up question if you don't mind. I understand that with blogdown you take Rmd to HTML, which Hugo can still process.

Before your package I was trying to use `rmarkdown::render("input.Rmd", rmarkdown::md_document(preserve_yaml = TRUE),...)` in the hope that I could add htmlwidgets via shortcodes but I'd always get thwarted because of < and >. 

Obviously it's great that I can use shortcodes in Rmd files using blogdown, so this might be a moot point now. Thanks again!

**Comment by lcolladotor** [comment]
@mwhitaker if you want an `.md` file use `knitr::render_markdown(TRUE)`. It'll work with the shortcodes.

**Comment by mwhitaker** [comment]
Thanks @lcolladotor but this is somehow not working for me. Maybe I am doing this wrong, but isn't `knitr::render_markdown(TRUE)` supposed to be used in a "setup" chunk in the Rmd file? I did that and then I used `rmarkdown::render("input.Rmd", rmarkdown::md_document(preserve_yaml = TRUE),...)` to render to .md. But I still get <> encoded.

**Comment by lcolladotor** [comment]
No, use it to render the `.Rmd` file directly to get a `.md` file that still have the shortcode intact. See https://github.com/lcolladotor/markdown-redcarpet.tmbundle/commit/f043c056ff620299843e9d8ea34144f478aa7965#diff-2c37ba0c676929fe1064b758c2fccf15R39

**Comment by mwhitaker** [comment]
thank you @lcolladotor - works great!

**Comment by yihui** [comment]
FYI I have provided `blogdown::shortcode()` in the current dev version that you can use in R Markdown documents to write shortcodes. See its help page for more info.

**Comment by EllaKaye** [comment]
Hi @yihui. Thanks for this! Small point, but I think there's a typo in the example on the help page for this. It gives `shortcode("tweet", 1234567)`, but when I tried that (with an appropriate tweet ID), it didn't work, but putting the ID in as a string does.

**Comment by yihui** [comment]
@EllaKaye Thanks! You are correct that the tweet ID should be quoted.

**Comment by pssguy** [comment]
Sorry to be a pain, but I'm missing something pretty basic here
The html output of an Rmarkdown document includes this

![shortcodeissue](https://user-images.githubusercontent.com/2775918/28503359-dd2a46f4-6fb9-11e7-8269-9a0262375a8f.PNG)

This seems to tally with the documentation

"Value
A character string wrapped in htmltools::HTML(); shortcode() returns a string of the form {{% name args %}},"

but I want to see the actual tweet

**Comment by yihui** [comment]
@pssguy Could you provide a reproducible example? `shortcode()` is a function specific to blogdown and Hugo. You cannot use it elsewhere.

**Comment by pssguy** [comment]
False Alarm, I think
It was after knitting that single page - so guess Hugo would not be involved, presumably
With serve_site() it worked OK - though on one occasion testing  I had to refresh the browser

**Comment by yihui** [comment]
The documentation says (https://bookdown.org/yihui/blogdown/workflow.html):

> Do not click the Knit button on the RStudio toolbar.

**Comment by andry-tino** [comment]
The very nice thing about using this approach in `.Rmd` files is that you can save the values of parameters inside variables. A very good scenario is when you want to have shortcodes for embedded content. For a generic shortcode, you will need to pass the embed string which might contain quotes in it. You can store that into a variable and pass it to the shortcode :)

**Comment by sagitaninta** [comment]
Hi everyone, I was just trying to render my .Rmd post that used the shortcode for twitter card and get the following message while running `blogdown::serve_site()`:

`Timed out creating content for page "D:\\Master\\Rcourse\\academic-kickstart-master\\content\\post\\2019-04-10-making-website-using-blogdown-and-hugo.html" (.Content will be empty). This is most likely a circular shortcode content loop that should be fixed. If this is just a shortcode calling a slow remote service, try to set "timeout=30000" (or higher, value is in milliseconds) in config.toml.`

I have tried to set the timeout parameter in config.toml... 
![image](https://user-images.githubusercontent.com/28356594/55937422-298af100-5c63-11e9-8915-9572483be576.png)
...but I still got post with empty content
![image](https://user-images.githubusercontent.com/28356594/55937666-be8dea00-5c63-11e9-997c-9bfc58f47f3e.png)

Only after I set my r chunks containing the `blogdown::shortcode()` functions to `eval=FALSE` I can see my content. But that is not the point, I want the twitter cards :cries: If this is because of the internet connection, the gif images from giphy is shown just fine.

If anyone here could help me, I would very much appreciate this. Thank you for this awesome package.

**Comment by andry-tino** [comment]
I think there might be a issue with function `blogdown::shortcode`. I posted it here: [Cannot use Markdown shortcodes in Blogdown's Rmd files
](https://stackoverflow.com/questions/55856904/cannot-use-markdown-shortcodes-in-blogdowns-rmd-files).

**Comment by giabaio** [comment]
I've experience this issue too. The simplest workaround is of course to embed the pure HTML code within the shortcode, which however defies the purpose of having the shortcode in the first place... Not sure what the issue is though.

**Comment by yihui** [comment]
I think Hugo has broken shortcodes of the form `{{% %}}` since v0.55. Several users have reported it: https://github.com/gohugoio/hugo/issues?q=is%3Aissue+shortcode+.inner+markdown+is%3Aclosed For example:

- https://github.com/gohugoio/hugo/issues/5870
- https://github.com/gohugoio/hugo/issues/5888
- https://github.com/gohugoio/hugo/issues/5900
- https://github.com/gohugoio/hugo/issues/5902
- https://discourse.gohugo.io/t/18315

From what I read from the Hugo main author's responses, it seems you'll have to try harder to convince him that this is a really bad breakage. I agree with @matteocontrini in https://github.com/gohugoio/hugo/issues/5870: now it seems `%` and `<>` don't have a difference.

**Comment by mikemc** [comment]
> The simplest workaround is of course to embed the pure HTML code within the shortcode, which however defies the purpose of having the shortcode in the first place - @giabaio 

I'm probably not fully following the discussion, but just wanted to point out that @yihui 's trick [mentioned above](https://github.com/rstudio/blogdown/issues/40#issuecomment-275134811) still works in Hugo 0.55 and seems to maintain the benefits of the shortcode.

**Comment by giabaio** [comment]
Thanks, @yihui and @mikemc for yours. In my specific case, the simplest thing to do is to use something like 
```
{{% alert note %}}
<b>Some bold text</b> <a href="someURL">here</a>
{{% /alert %}}
```
because not all  of my pages are `Rmd` (as legacy of how I've built up the whole website. Plus I don't need all the pages to be `Rmd`, as in many cases, pure `md` would do just fine and it's probably quicker to then render?).

So that's what I meant to embed pure `HTML` in the shortcode, whereas before HUGO broke them, I could use simple
```
{{% alert note %}}
**Some bold text** [here](someURL)
{{% /alert %}}
```
which is a bit annoying (as the `blogdown` version is easier and more in line with the rest of the page) --- but ultimately not the end of the world (for me)...

Does this make sense?

**Comment by yihui** [comment]
@giabaio Did you read my SO answer mentioned above? https://stackoverflow.com/a/55928449/559676

**Comment by giabaio** [comment]
Hi @yihui. I hadn't --- thanks for this!

**Comment by jcolomb** [comment]
Hello,

Is there a way to keep markdown synthax inside the shortcode() working, or do I have to go to a html synthax. In the following example, the link `[BIDS data format](https://bids.neuroimaging.io/)` is not a link anymore:

```
blogdown::shortcode ("info", title = "summary",
           .content = "I am working on [BIDS data format](https://bids.neuroimaging.io/). .")
```

**Comment by yihui** [comment]
@jcolomb Like @giabaio, you do not seem to have read my reply above: https://github.com/rstudio/blogdown/issues/40#issuecomment-488107934.

**Comment by giabaio** [comment]
Sorry! I did read this but forgot to reply! Many apologies!
Thanks Yihui!


On Tue, 10 Dec 2019, 19:12 Yihui Xie, <notifications@github.com<mailto:notifications@github.com>> wrote:

@jcolomb<https://github.com/jcolomb> Like @giabaio<https://github.com/giabaio>, you do not seem to have read my reply above: #40 (comment)<https://github.com/rstudio/blogdown/issues/40#issuecomment-488107934>.

**Comment by jcolomb** [comment]
sorry, I did not realise it was working before the hugo update,thanks for replying anyway!

**Comment by yihui** [comment]
In my opinion, it was a bad change in Hugo and extremely confusing.  I don't think it is your fault.

I'm going to lock this issue to make it easier for future visitors of this thread to find the solution at https://github.com/rstudio/blogdown/issues/40#issuecomment-488107934. If you still have issues related to shortcodes, please feel free to open a new issue. Thanks!
